### ISSUE TITLE
overlay network exhaustion

### ISSUE THREAD

**Issue by chrisbecke** [moby/moby#43052]
**Description**

I have a production docker swarm that uses traefik for ingress routing. Recently, services fail to deploy, getting stuck with a current state of "New (Creating)" `docker service inspect` does not show why the service is not deploying, neither does `docker inspect <task-id>`, and `docker service logs is empty` as the service has not even been scheduled on a node yet.
Only when visiting the manager node is the following error found via `journalctl -xfu docker.service`

```bash
Nov 30 11:09:05 lab717 dockerd[35784]: time="2021-11-30T11:09:05.639670358+02:00" level=error msg="Failed allocation during update of service o7ketufqiyy9jgbjt26irqimo" error="could not find an available IP while allocating VIP" module=node node.id=n44h9ubgs7r7i09z97nwd09iv
```
**Steps to reproduce the issue:**
1. Have a swarm using traefik for ingress routing. 
2. Attach ~100 services or containers to the global traefik_public overlay network.

**Describe the results you received:**

- vips are exhausted after only ~100 services are attached even though there should be ~250 ips are available.
- no error is visible to a docker user. Only sys-admins can access the reason services are stuck in the "New" state.

**Describe the results you expected:**

- An error attached to the task / service that can be accessed via `docker xxx inspect`.

**Discussion of mitigations:**

Of course, nevermind the inability to diagnose the exhaustion, mitigations appear largely locked out:

- #30820 indicates my plan of making the traefik network bigger is bound to fail. Is this seriously still a thing?

- Given networks should not be made larger than /24 - then are there any potential ways to get more useable ips per network? There are no documents describing how many vips docker creates, what a vip is, how to enumerate them, or how to monitor and alert on available address space in a network.

**Output of `docker version`:**

```
Client:
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:55:20 2021
 OS/Arch:           darwin/amd64
 Context:           lab727
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:16:13 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    lab717
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  compose: Docker Compose (Docker Inc., v2.0.0-rc.3)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 30
  Running: 30
  Paused: 0
  Stopped: 0
 Images: 75
 Server Version: 20.10.2
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: local
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: n44h9ubgs7r7i09z97nwd09iv
  Is Manager: true
  ClusterID: 7rcpc2tmbrgqn58me57ruoazy
  Managers: 3
  Nodes: 14
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.1.7.90
  Manager Addresses:
   10.1.7.104:2377
   10.1.7.116:2377
   10.1.7.90:2377
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1062.12.1.el7.x86_64
 Operating System: Red Hat Enterprise Linux Server 7.9 (Maipo)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.62GiB
 Name: lab717
 ID: JZ3A:GO4W:O2F5:KBTP:YLMM:4ZTJ:XVDD:K45J:KSJX:I6JY:WCAC:VNH2
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Registry Mirrors:
  https://docker-hub-proxy.unreal.mgsops.net/
  http://digit-cache.unreal.mgsops.net/
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

vmWare vCloud

**Comment by emilioSp** [comment]
I have your exact issue with traefik and docker swarm. Have you found a solution? @chrisbecke 

cc: @disheng

**Comment by chrisbecke** [comment]
@emilioSp 

"solving" this was tricky:

I had to restart all my docker nodes after editing daemon.json so that docker didn't use the entire 172.16.0.0/12 cidr for its bridge networking. This freed up some ranges that were now totally unassignable by docker. Because docker is stupid and if you manually create a network range using the bridge and/or overlay drivers it ignores your manual allocation and will auto create a network in that range if/when it gets there.

With an unassigned range in hand I manually created a bigger traefik network: /20 in size iirc. #30820 is a thing however so we need to mitigate for that below).

This was actually more complicated. As I had a running swarm, I created first a temporary holding network and scripted the addition of a new network to all containers and services to the new network that existed on the old traefik. I could then script the removal of everything from `traefik` and then delete traefik. Then I could create the new traefik and add `traefik` to everything that was in the holding network. Then I could delete the holding network from everything and delete it. A mission.

To avoid falling afoul of #30820 we did an audit of all the services that were attached to traefik and ensured that they were all flipped to `endpoint_mode: dnsrr`. For some services this was not appropriate. As long as the set of exceptions is kept small #30820 won't be triggered. Otherwise if you can't set dnsrr on a service because it is also being used as an api or something, we created a socat proxy, and moved the traefik labels to that.

tl;dr
1. Get a reserved /20 range in the reserved IP space, which is hard because docker grabs everything by default.
2. Create a new traefik network and add it to all the services and containers that were on the old /24 traefik network.
3. Make sure those services are using `endpoint-mode: dnsrr`, as `endpoint-mode: vip` does not scale past /24 on overlay networks. Use proxies where a simple swap isn't approprioate.

**Comment by dusatvoj** [comment]
@chrisbecke it's not recommended to run larger than /24 networks @ swarm https://github.com/moby/moby/issues/37338#issuecomment-859203568

**Comment by chrisbecke** [comment]
@dusatvoj the /24 advice is related to this issue #30820 which ... was fixed in .22 iirc. #37338 refers to leaked memory when containers reallocate IPs and that can be triggered with a small number of containers that continually restart. The bane of my life. And then there is a new issue that give a hard cap of 4096? IPs on overlay networks that's still open so the practical current limit is /20 I think (and fix any services with tasks that are stuck in restart loops),

**Comment by dusatvoj** [comment]
@chrisbecke , thank's for explaining. :pray:
